In [1]:
# importing libraries
import pandas as pd
import numpy as np
import seaborn as sns #visualisation
import matplotlib.pyplot as plt #visualisation
%matplotlib inline 
sns.set(color_codes=True)
from pandasql import sqldf

In [2]:
#  loading the dataset
df = pd.read_csv('../asset/hotel_bookings.csv')

# Understanding the dataset

First 5 rows

In [3]:

df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

Last 5 rows

In [4]:
df.tail()

hotel  is_canceled  lead_time  arrival_date_year  \
119385  City Hotel            0         23               2017   
119386  City Hotel            0        102               2017   
119387  City Hotel            0         34               2017   
119388  City Hotel            0        109               2017   
119389  City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
119385                            0                          0   
119386                            0                          2   
119387                            0                          4   
119388                            0                          0   
119389                            0                          2   

        reservation_status reservation_status_date  
119385           Check-Out              2017-09-06  
119386           Check-Out              2017-09-07  
119387           Check-Out              2017-09-07  
119388           Check-Out              2017-09-07  
119389           Check-Out              2017-09-07  

[5 rows x 32 columns]

Total number of rows and columns

In [5]:
print('Total number of rows:',df.shape[0],'\nTotal number of columns:',df.shape[1])



Total number of rows: 119390 
Total number of columns: 32


Data type of each column

In [6]:
df.dtypes

hotel                              object
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
market_segment                     object
distribution_channel               object
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                 object
assigned_room_type                 object
booking_changes                     int64
deposit_type                       object
agent                             

Other useful information about the dataset

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

# Data Cleaning

### Handling null values

Finding the null values

In [8]:
print(df.isnull().sum())

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

Replacing null values with appropriate values according to the respective column

In [9]:
# replacing null values in children with 0
df["children"].fillna(0, inplace = True)


# replacing null values in country with No Country
df["country"].fillna("No Country", inplace = True)


# replacing null values in agent with 0
df["agent"].fillna(0, inplace = True)


# replacing null values in company with No Country
df["company"].fillna("No Company", inplace = True)

In [10]:
print(df.isnull().sum())

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                         

In [11]:
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent     company days_in_waiting_list customer_type   adr  \
0    0.0  No Company                    0     Transient   0.0   
1    0.0  No Company                    0     Transient   0.0   
2    0.0  No Company                    0     Transient  75.0   
3  304.0  No Company                    0     Transient  75.0   
4  240.0  No Company                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

### Handling duplicates

Finding duplicates

In [12]:
duplicate_rows_df = df[df.duplicated()]
print('Number of duplicate rows:', duplicate_rows_df.shape[0])

Number of duplicate rows: 31994


In [13]:
# before removing duplicate rows
df.count() 

hotel                             119390
is_canceled                       119390
lead_time                         119390
arrival_date_year                 119390
arrival_date_month                119390
arrival_date_week_number          119390
arrival_date_day_of_month         119390
stays_in_weekend_nights           119390
stays_in_week_nights              119390
adults                            119390
children                          119390
babies                            119390
meal                              119390
country                           119390
market_segment                    119390
distribution_channel              119390
is_repeated_guest                 119390
previous_cancellations            119390
previous_bookings_not_canceled    119390
reserved_room_type                119390
assigned_room_type                119390
booking_changes                   119390
deposit_type                      119390
agent                             119390
company         

In [14]:
# removing duplicate rows
df = df.drop_duplicates()

# after removing duplicate rows
df.count() 

hotel                             87396
is_canceled                       87396
lead_time                         87396
arrival_date_year                 87396
arrival_date_month                87396
arrival_date_week_number          87396
arrival_date_day_of_month         87396
stays_in_weekend_nights           87396
stays_in_week_nights              87396
adults                            87396
children                          87396
babies                            87396
meal                              87396
country                           87396
market_segment                    87396
distribution_channel              87396
is_repeated_guest                 87396
previous_cancellations            87396
previous_bookings_not_canceled    87396
reserved_room_type                87396
assigned_room_type                87396
booking_changes                   87396
deposit_type                      87396
agent                             87396
company                           87396


#  Exploratory analysis 

In [15]:
pysqldf = lambda q: sqldf(q, globals())

In [16]:
q = """select hotel 
       from df 
       LIMIT 100;"""

In [17]:
hotel = pysqldf(q)
hotel

hotel
0   Resort Hotel
1   Resort Hotel
2   Resort Hotel
3   Resort Hotel
4   Resort Hotel
..           ...
95  Resort Hotel
96  Resort Hotel
97  Resort Hotel
98  Resort Hotel
99  Resort Hotel

[100 rows x 1 columns]

### Hotel

Total Number of hotels in LimeHome's Program

In [18]:
q = """SELECT COUNT(DISTINCT hotel) as Number_of_hotels_in_LimeHomes_Program
        from df;"""


hotel = pysqldf(q)
hotel

Number_of_hotels_in_LimeHomes_Program
0                                      2

Most/least visited hotel

In [19]:
q = """SELECT 
        COUNT(CASE when hotel = 'Resort Hotel' then 'yes' END) as Resort_Hotel_number_of_visits,
        COUNT(CASE when hotel= 'City Hotel' then 'yes' END) as City_Hotel_number_of_visits
        from df
        where reservation_status = 'Check-Out' """

hotel = pysqldf(q)
hotel

Resort_Hotel_number_of_visits  City_Hotel_number_of_visits
0                          25992                        37379

### Year and Month

Year with the most visits

In [20]:
q = """SELECT arrival_date_year,
        count (arrival_date_year) as value_occurrence_year
        from df
        group by arrival_date_year
        order by value_occurrence_year desc"""

year_month = pysqldf(q)
year_month

arrival_date_year  value_occurrence_year
0               2016                  42391
1               2017                  31692
2               2015                  13313

Month with the most visit

In [21]:
q = """SELECT arrival_date_month,
        count (arrival_date_month) as value_occurrence
        from df
        group by arrival_date_month
        order by value_occurrence desc"""

year_month = pysqldf(q)
year_month

arrival_date_month  value_occurrence
0              August             11257
1                July             10057
2                 May              8355
3               April              7908
4                June              7765
5               March              7513
6             October              6934
7           September              6690
8            February              6098
9            December              5131
10           November              4995
11            January              4693

Finding out the months with most visits in the year 2016

In [22]:
q = """SELECT arrival_date_month,
        count (arrival_date_month) as value_occurrence
        from df
        where arrival_date_year = '2016'
        group by arrival_date_month
        order by value_occurrence desc"""

year_month = pysqldf(q)
year_month




arrival_date_month  value_occurrence
0              August              4423
1             October              4234
2           September              3851
3                July              3851
4               March              3831
5                 May              3780
6               April              3770
7                June              3517
8            November              3330
9            December              3149
10           February              2806
11            January              1849

### Country

Number of countries that LimeHome Program exist

In [23]:
q = """SELECT COUNT(DISTINCT country) as Number_of_countries_in_LimeHomes_Program
        from df;"""


country = pysqldf(q)
country

Number_of_countries_in_LimeHomes_Program
0                                       178

Most/least visited countries

In [24]:
q = """SELECT country,
        count (country) as value_occurrence
        from df
        group by country
        order by value_occurrence desc"""

country = pysqldf(q)
country

country  value_occurrence
0       PRT             27453
1       GBR             10433
2       FRA              8837
3       ESP              7252
4       DEU              5387
..      ...               ...
173     BFA                 1
174     BDI                 1
175     ATF                 1
176     ASM                 1
177     AIA                 1

[178 rows x 2 columns]

Checking the month in which we have the most visits in PRT (Portugal).

In [25]:
q = """SELECT arrival_date_month,
        count (arrival_date_month) as value_occurrence
        from df
        where country = 'PRT' 
        group by arrival_date_month
        order by value_occurrence desc"""

country = pysqldf(q)
country

arrival_date_month  value_occurrence
0              August              3629
1                July              3260
2               March              2336
3            February              2217
4           September              2196
5                June              2154
6             October              2127
7                 May              2081
8            December              1983
9               April              1974
10            January              1757
11           November              1739

 ### Meal 

Most demanded meal

In [26]:
q = """SELECT meal,
        count (meal) as value_occurrence
        from df
        group by meal
        order by value_occurrence desc"""

meal = pysqldf(q)
meal

meal  value_occurrence
0         BB             67978
1         SC              9481
2         HB              9085
3  Undefined               492
4         FB               360

Most demanded meal in PRT

In [27]:
q = """SELECT meal,
        count (meal) as value_occurrence
        from df
        where country = 'PRT' 
        group by meal
        order by value_occurrence desc"""

meal = pysqldf(q)
meal

meal  value_occurrence
0         BB             22283
1         HB              3160
2         SC              1484
3  Undefined               267
4         FB               259

### distribution_channel

Most used distribution channel

In [28]:
q = """SELECT distribution_channel,
        count (distribution_channel) as value_occurrence
        from df
        group by distribution_channel
        order by value_occurrence desc"""

distribution_channel = pysqldf(q)
distribution_channel

distribution_channel  value_occurrence
0                TA/TO             69141
1               Direct             12988
2            Corporate              5081
3                  GDS               181
4            Undefined                 5

### customer_type

Available customer types 

In [29]:
q = """SELECT customer_type,
        count (customer_type) as value_occurrence
        from df
        group by customer_type
        order by value_occurrence desc"""


customer_type = pysqldf(q)
customer_type

customer_type  value_occurrence
0        Transient             71986
1  Transient-Party             11727
2         Contract              3139
3            Group               544

Average number of booking changes made by Transient customer types 

In [30]:
q = """SELECT 
       AVG(booking_changes),  MAX(booking_changes)
        from df
        where (customer_type='Transient' and booking_changes >=1)
        """

customer_type = pysqldf(q)
customer_type

AVG(booking_changes)  MAX(booking_changes)
0              1.466921                    20

Average number of previous cancellations made by Transient customer type

In [ ]:

q = """SELECT avg(previous_cancellations), max(previous_cancellations)
        from df
        where (customer_type='Transient' and previous_cancellations>=1 )

    """


previous_cancellations = pysqldf(q)
previous_cancellations

Transient customer type average number of children 

In [ ]:
q = """SELECT 
       AVG(children),  MAX(children)
        from df
        where (customer_type='Transient' and is_canceled =0)
        """

customer_type = pysqldf(q)
customer_type

### Reserved Room

Most/least reserve room type

In [ ]:
q = """SELECT reserved_room_type,
        count (reserved_room_type) as value_occurrence
        from df
        group by reserved_room_type
        order by value_occurrence desc"""

room = pysqldf(q)
room

Most/least reserve room type by Transient customer type

In [ ]:
q = """SELECT reserved_room_type,
        count (reserved_room_type) as value_occurrence
        from df
        where (customer_type='Transient' and is_canceled =0)
        group by reserved_room_type
        order by value_occurrence desc"""

room = pysqldf(q)
room

### Parking Spaces 

The number of parking spaces often required by all customers

In [ ]:
q = """SELECT required_car_parking_spaces,
        count (required_car_parking_spaces) as value_occurrence
        from df
        group by required_car_parking_spaces
        order by value_occurrence desc"""

parking = pysqldf(q)
parking

The number of parking spaces often required by Transient customer type

In [ ]:
q = """SELECT required_car_parking_spaces,
        count (required_car_parking_spaces) as value_occurrence
        from df
        where (customer_type='Transient' and is_canceled =0)
        group by required_car_parking_spaces
        order by value_occurrence desc"""

parking = pysqldf(q)
parking

# Hypotheses

## Hypothesis 1

As explored, 2016 was the year with the most touristc travels and Portugal was the most visited country. Even in 2016, Portugal was the most visited country, see results below.

In [ ]:
q = """SELECT country,
        count (country) as value_occurrence
        from df
        where arrival_date_year = '2016'
        group by country
        order by value_occurrence desc"""

country = pysqldf(q)
country

My hypothesis to why Portugal had the most visit, even more that what Britian had was because of the temperature in Portugal was higher, especially in the summer. 

In addition, the Portugal News (2017) reported Portugal broke the weather records in June, July, August and September in 2016. Temperatures reaching record highs, with exceptionally low precipitation. 



References: 

Holiday Weather, Algarve, Portugal: Annual Weather Averages. https://www.holiday-weather.com/algarve/averages/

Holiday Weather, London, United Kingdom: Annual Weather Averages. https://www.holiday-weather.com/london/averages/

The Portugal News, Portugal’s €7bn climate change bill, 2017. https://www.theportugalnews.com/news/portugals-7bn-climate-change-bill/40860#:~:text=In%20terms%20of%20the%20effects,by%20temperatures%20reached%20in%201989. 

Checking the month with the highest visit in Portugal 

In [ ]:
q = """SELECT arrival_date_month,
        count (arrival_date_month) as value_occurrence
        from df
        where ( country = 'PRT' )
        group by arrival_date_month
        order by value_occurrence desc"""

country = pysqldf(q)
country

Although Spain close to Portugal and even has higher temperature than Portugal, Portugal had more visitors than Spain. There are a lot of reasons to why this is the case. Some are: hotel prices in both countries, proximity to the sea and beaches, population density of the tourist cities, tourism policies etc. 


Reference:

Almeida García, Fernando & Macias, Rafael & Balbuena, Antonia. (2015). Tourist Spaces and Tourism Policy in Spain and Portugal. International Journal of Culture Tourism and Hospitality Research. 10. 235-249. 10.1108/S1871-317320150000010017. 

## Hypothesis 2



With over 67,978 (restrictly BB) room reservations including breakfast , it is of no surprised why this is the case. For decades there has been a lot of articles proving why breakfast is the most important meal of the day. The convenience and benefit of having breakfast even when you are far from home in a hotel is often worth the additional fee that hotels or Airbnb charge for breakfast.  


BB (Bed and Breakfast) —— Breakfast s included. 

SC (Self Catering) —— No meals are included; however, your accommodation will be provided with catering facilities for you to cook light meals.

HB (Half Board) —— Breakfast and evening meals are included. In some cases, you can choose to receive lunch instead of breakfast – the hotel will confirm this on arrival.



Reference:

Charles Spence,
Breakfast: The most important meal of the day?,
International Journal of Gastronomy and Food Science,
Volume 8,
2017,
Pages 1-6,
ISSN 1878-450X,
https://doi.org/10.1016/j.ijgfs.2017.01.003. 


Benton et al., 2001
D. Benton, O. Slater, R.T. Donohoe
The influence of breakfast and a snack on psychological functioning
Physiol. Behav., 74 (2001), pp. 559-571


# Seasonality

My first hypothesis already indicates that there is seasonality. There is more reservations/travels/touristic activaites in the summer that any other season. The least reservations were observed in the autumn and winter. 
Also to be noted is that reservations starts increasing in spring as the temperature gets warmer, and then peaks in the summer. 

In [ ]:
q = """SELECT arrival_date_month,
        count (arrival_date_month) as value_occurrence
        from df
        group by arrival_date_month
        order by value_occurrence desc"""

country = pysqldf(q)
country

# Correlation between lead time and cancellation

In [ ]:
df['is_canceled'].corr(df['lead_time'])

## Running a simple logistic regression 

### Defining the features and labels

Lead Time: In a hotel system, lead time is the length of time between which a customer makes a reservation and their actual arrival. This will be independent variable  or feature (X) in the machine learning classification model that will be built.

Cancelation: This tells us whether a customer has canceled their reservation or not. This will be our dependent variable or label (y).
0 —— means the customer has NOT canceled their reservation
1 —— means the customer has CANCELED their reservation 


In [ ]:


X= df['lead_time'].values
X = X.reshape(-1, 1)
X

In [ ]:
y = df['is_canceled'].values
y

### Data split

In [ ]:

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [ ]:
X_test

### Feature scaling

In [ ]:



from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
X_test

### Training the model

In [ ]:

from sklearn.linear_model import LogisticRegression # import the  Logistic Regression class from its lib
classifier = LogisticRegression (random_state=0)


# from sklearn.svm import SVC 
# classifier = SVC (kernel = 'linear',  random_state = 0, )


classifier.fit(X_train,y_train)

### Predicting the test result

In [ ]:

y_pred = classifier.predict(X_test)
y_pred

### Model evaluation

In [ ]:

from sklearn.metrics import confusion_matrix, accuracy_score # this function compares the predicted results and the actual result, 
cm = confusion_matrix (y_test, y_pred) 

print("Prediction Accuracy:",accuracy_score(y_test, y_pred), '\n\nConfusion Matrix:\n',cm)

In [ ]:
from sklearn.metrics import plot_confusion_matrix
plot_confusion_matrix(classifier, X_test, y_test)  
plt.show()

### New Predictions

In [ ]:
lead_time = 215
X_new = [[lead_time]]

# applying the StandardScaler 
X_new = sc_X.transform(X_new)



y_new = classifier.predict(X_new)
y_new[0]

if y_new[0] == 0:
    print('This customer will NOT cancel his/her booking.\nPredicted class = ',y_new[0])
elif y_new[0] == 1:
    print('This customer WILL cancel his/her booking.\nPredicted class = ',y_new[0])
else:
    print('Something is not right.')

### Model Summary

The higher the lead time the most probable that the customer will cancel their booking. 

This is very logical, because as time goes on after booking thier reservation, and before setting out for the journey, a lot of things can happen to make them cancel thier booking, example, finding a cheaper option or a more conducive room in another hotel or airbnb.

To prevent this, most hotel managements have some insurancy policy — Lead Time Restrictions. 
Some require the customer to pay full or half of the fee immediately after booking. 
Some give thier customers a specified time length before the set arrival date that the customer can freely cancel their  reservation without cancelation charges.



